File: EXAMEN/EXAMEN 2/civitasRuby/lib/vista_textual.rb
#encoding:utf-8
require_relative 'operaciones_juego'
require_relative 'salidas_carcel'
require_relative 'respuestas'
require_relative 'gestiones_inmobiliarias'
require 'io/console'

module Juego_texto

  class Vista_textual
  include Civitas
    
    attr_reader :ipropiedad, :igestion

    def mostrar_estado(estado)
      puts estado
    end

    def pausa
      print "Pulsa una tecla "
      STDIN.getch
      print "\n"
    end

    def lee_entero(max,msg1,msg2)
      ok = false
      begin
        print msg1
        cadena = gets.chomp
        begin
          if (cadena =~ /\A\d+\Z/)
            numero = cadena.to_i
            ok = true
          else
            raise IOError
          end
        rescue IOError
          puts msg2
        end
        if (ok)
          if (numero >= max)
            ok = false
          end
        end
      end while (!ok)

      return numero
    end


    def menu(titulo,lista)
      tab = "  "
      puts titulo
      index = 0
      lista.each { |l|
        puts tab+index.to_s+"-"+l
        index += 1
      }

      opcion = lee_entero(lista.length,
                          "\n"+tab+"Elige una opción: ",
                          tab+"Valor erróneo")
      return opcion
    end

    def salir_carcel
      opcion = menu("Elige la forma para intentar salir de la carcel", 
                    Array(["Pagando","Tirando el dado"]) )
      return (Lista_salidas_carcel[opcion]) 
    end

    def comprar
      opcion = menu("¿Deseas comprar #{@juego_model.get_casilla_actual.nombre}?",
                    Array(["Si", "No"]))
      return(Lista_respuestas[opcion])
    end

    def gestionar
      @igestion = menu("¿Qué gestión desea realizar?",
        Array(["Vender", "Hipotecar", "Cancelar hipoteca", "Construir casa", "Construir Hotel", "Hacer propiedad sostenible" , "Terminar"]))
      if @igestion != 6
        nombres_propiedades = Array.new
        for i in 0..(@juego_model.get_jugador_actual.propiedades.size - 1)
          nombres_propiedades << @juego_model.get_jugador_actual.propiedades[i].to_s
        end
        @ipropiedad =  menu("¿Sobre qué propiedad?", nombres_propiedades)
       end
    end

    def mostrar_siguiente_operacion(operacion)
      puts "La siguiente operacion que se va a realizar es #{operacion}"
    end

    def mostrar_eventos
      while (Diario.instance.eventos_pendientes)
          puts Diario.instance.leer_evento
      end

    end

    def set_civitas_juego(civitas)
         @juego_model=civitas
         self.actualizar_vista
    end

    def actualizar_vista
      puts "Jugador actual: #{@juego_model.info_jugador_texto}"
      puts "Casilla actual: #{@juego_model.get_casilla_actual.to_s}"
    end

    
  end

end

File: Practica_3/P1/civitasJava/src/juegoTexto/VistaTextual.java
package juegoTexto;

import civitas.Barrio;
import civitas.CivitasJuego;
import civitas.Diario;
import civitas.OperacionesJuego;
import civitas.SalidasCarcel;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;
import civitas.Casilla;
import civitas.CivitasJuego;
import civitas.Jugador;
import civitas.OperacionesJuego;
import civitas.SalidasCarcel;
import civitas.TituloPropiedad;
import civitas.Respuestas; 
import civitas.Tablero;
 
 class VistaTextual {
  
  CivitasJuego juegoModel; 
  int iGestion=-1;
  int iPropiedad=-1;
  private static String separador = "=====================";
  
  private Scanner in;
  
   VistaTextual () {
    in = new Scanner (System.in);
  }
  
  void mostrarEstado(String estado) {
    System.out.println (estado);
  }
              
  void pausa() {
    System.out.print ("Pulsa una tecla ");
    in.nextLine();
  }

  int leeEntero (int max, String msg1, String msg2) {
    Boolean ok;
    String cadena;
    int numero = -1;
    do {
      System.out.print (msg1);
      cadena = in.nextLine();
      try {  
        numero = Integer.parseInt(cadena);
        ok = true;
      } catch (NumberFormatException e) { // No se ha introducido un entero
        System.out.println (msg2);
        ok = false;  
      }
      if (ok && (numero < 0 || numero >= max)) {
        System.out.println (msg2);
        ok = false;
      }
    } while (!ok);

    return numero;
  }

  int menu (String titulo, ArrayList<String> lista) {
    String tab = "  ";
    int opcion;
    System.out.println (titulo);
    for (int i = 0; i < lista.size(); i++) {
      System.out.println (tab+i+"-"+lista.get(i));
    }

    opcion = leeEntero(lista.size(),
                          "\n"+tab+"Elige una opción: ",
                          tab+"Valor erróneo");
    return opcion;
  }

  SalidasCarcel salirCarcel() {
    int opcion = menu ("Elige la forma para intentar salir de la carcel",
      new ArrayList<> (Arrays.asList("Pagando","Tirando el dado")));
    return (SalidasCarcel.values()[opcion]);
  }

//  Respuestas comprar() {
//    int opcion = menu ("¿Deseas comprar " + juegoModel.getCasillaActual().getNombre() + "?",
//      new ArrayList<> (Arrays.asList("Si", "No")));
//    return (Respuestas.values()[opcion]);
//  }

  void gestionar () {
    iGestion= menu("¿Qué gestión quieres realizar?",
      new ArrayList<> (Arrays.asList("Vender", "Hipotecar", "Cancelar hipoteca", "Construir casa", "Construir hotel", "Terminar")));

      ArrayList<String> nombresPropiedades = new ArrayList();

    for(int i = 0; i < juegoModel.getJugadorActual().getPropiedades().size(); i++)
      nombresPropiedades.add(juegoModel.getJugadorActual().getPropiedades().get(i).toString());

    if (iGestion != 5) // Si la gestion no es terminar
        iPropiedad = menu("¿Sobre qué propiedad?", nombresPropiedades);
    
  }
  
  public int getGestion(){
    return iGestion;
  }
  
  public int getPropiedad(){
    return iPropiedad;
  }
    

  void mostrarSiguienteOperacion(OperacionesJuego operacion) {
    System.out.println("La siguiente operación que se va a realizar es " + operacion);
  }


  void mostrarEventos() {
    while(Diario.getInstance().eventosPendientes())
      System.out.println(Diario.getInstance().leerEvento());
  }
  
  public void setCivitasJuego(CivitasJuego civitas){ 
        juegoModel=civitas;
        this.actualizarVista();
  }
  
  void actualizarVista(){
    System.out.println("Jugador actual: " + juegoModel.infoJugadorTexto());
    System.out.println("Casilla actual: " + juegoModel.getCasillaActual().toString());
  } 
  
  //Nuevo
  Respuestas comprar(){
      
      int njug = juegoModel.getIndiceJugadorActual();
      Jugador jugadorActual = juegoModel.getJugadorActual();
      int numCasilla = jugadorActual.getNumCasilaActual();
      Tablero tablero = Tablero.getInstance();
      ArrayList <Casilla> casillas = tablero.getCasillas();
      Casilla casilla = tablero.getCasilla(numCasilla);
      TituloPropiedad titulo = casilla.getTituloPropiedad();
      Barrio barrio = titulo.getBarrio();
      String estados;
      int opcion = menu (barrio.getEstadoInmuebles(),
      new ArrayList<> (Arrays.asList("Si","No")));
      return Respuestas.values()[opcion];
  }
  
}
